libbpf-rs/src/ringbuf.rs

```diff
@@ -6,82 +6,128 @@ use std::time::Duration;
-            if err != 0 {
-                return Err(Error::System(err as i32));
+        for (fd, callback) in self.fd_callbacks.clone() {
+            let sample_cb_ptr = Box::into_raw(callback);
+            if ptr.is_null() {
+                // Allocate a new ringbuf manager and add a ringbuf to it
+                ptr = unsafe {
+                    libbpf_sys::ring_buffer__new(
+                        fd,
+                        c_sample_cb,
+                        sample_cb_ptr as *mut _,
+                        std::ptr::null_mut(),
+                    )
+                };
+
+                // Handle errors
+                let err = unsafe { libbpf_sys::libbpf_get_error(ptr as *const _) };
+                if err != 0 {
+                    return Err(Error::System(err as i32));
+                }
+            } else {
+                // Add a ringbuf to the existing ringbuf manager
+                let err = unsafe {
+                    libbpf_sys::ring_buffer__add(ptr, fd, c_sample_cb, sample_cb_ptr as *mut _)
+                };
+
+                // Handle errors
+                if err != 0 {
+                    return Err(Error::System(err as i32));
+                }
             }
+
+            ringbufs.push(unsafe { Box::from_raw(sample_cb_ptr) });
         }
 
-        self.ringbufs.push(unsafe { Box::from_raw(sample_cb_ptr) });
+        if ptr.is_null() {
+            return Err(Error::InvalidInput(
+                "You must add at least one ring buffer map and callback before building".into(),
+            ));
+        }
 
-        Ok(self)
+        Ok(RingBuffer {
+            ptr,
+            _ringbufs: ringbufs,
+        })
     }
 
+    unsafe extern "C" fn call_sample_cb(ctx: *mut c_void, data: *mut c_void, size: u64) -> i32 {
+        let callback_ptr = ctx as *mut Callback;
+        let callback = &mut *callback_ptr;
+
+        callback(slice::from_raw_parts(data as *const u8, size as usize))
+    }
+}
+
+/// The canonical interface for managing a collection of `ringbuf` maps.
+///
+/// `ringbuf`s are a special kind of [`Map`], used to transfer data between
+/// [`Program`]s and userspace.  As of Linux 5.8, the `ringbuf` map is now
+/// preferred over the `perf buffer`.
+pub struct RingBuffer<F>
+where
+    F: FnMut(&[u8]) -> i32 + 'static,
+{
+    ptr: *mut libbpf_sys::ring_buffer,
+    _ringbufs: Vec<Box<F>>,
+}
+
+impl<F> RingBuffer<F>
+where
+    F: FnMut(&[u8]) -> i32 + 'static,
+{
     /// Poll from all open ring buffers, calling the registered callback for
     /// each one. Polls continually until we either run out of events to consume
     /// or `timeout` is reached.
     pub fn poll(&self, timeout: Duration) -> Result<()> {
         if self.ptr.is_null() {
-            return Err(Error::InvalidInput(
-                "You must add at least one ring buffer before polling".into(),
-            ));
+            return Err(Error::Internal("Null ringbuf pointer".into()));
```

Comment:
We error checked this in the builder so this can be replaced with an `assert!` or removed entirely (I prefer the assert!())

Reply:
Sure